### ISSUE TITLE
docker checkpoint create should not always store on host but also in containers

### ISSUE THREAD

**Issue by marc-christian-schulze** [docker/for-linux#233]
* [] This is a bug report
* [x] This is a feature request
* [x] I searched existing issues before opening this one

### Expected 
When creating a checkpoint and specifying a checkpoint directory I would expect the smae behaviour as with the docker cp command. If run on thehost the given destination path shall be relative to the hosts' file system and if run inside of a container the destination path should be relative to the container's gilesystem.

### Actual behavior
Creating a checkpoint with a specific checkpoint directory specify will write the checkpoint files always realtive to the host file system even if run from within a container. I would expect the docker checkpoint create --checkpoint-dir= command to behave like docker cp

### Steps to reproduce the behavior
```
host$ docker run some_image
host$ docker ps 
 ..... my_container
host$ docker run -v /var/run/docker.sock:/var/run/docker.sock -ti --rm docker
docker$ docker checkpoint create --checkpoint-dir=/test my_container some_checkpoint
docker$ ls /test
```

**Output of `docker version`:**

```
$ docker --version
Docker version 17.05.0-ce, build 89658b
```

**Output of `docker info`:**

```
$ docker info
Containers: 148
 Running: 0
 Paused: 0
 Stopped: 148
Images: 2225
Server Version: 17.05.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 2805
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9048e5e50717ea4497b757314bad98ea3763c145
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-112-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 6
Total Memory: 7.794GiB
Name: azrael-GA-990FXA-UD3
ID: HDRT:NZN2:C2EF:4P3U:J2G6:XZ4R:VE4J:HGQS:EXTF:UX4S:JBEC:XCMX
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.)**

Physical Ubuntu 16.04 Environment

**Comment by dnephin** [comment]
Any time you use expose the docker socket to a container you will have this problem.

It sounds like the `checkpoint create` is writing the checkpoint on the daemon host instead of the client, and you would expect it to send the data to the client and write it to the "client host".

**Comment by marc-christian-schulze** [comment]
Yes, that is what I observe when using docker cp. It streams the data through the docker socket to the client and stores the content in the local file system which also works inside of containers. Therefore, I would expect docker checkpoint doing the same instead of letting the daemon directly write the checkpoint to the host's filesystem.

**Comment by marc-christian-schulze** [comment]
I checked the implementation and found the cause for this behaviour.
While the `docker cp` command actively retrieves the desired content through a reader...
https://github.com/docker/docker-ce/blob/master/components/engine/client/container_copy.go#L63
... the `docker checkpoint` command simply submits the request for creating a checkpoint to the daemon
https://github.com/docker/docker-ce/blob/master/components/engine/client/checkpoint_create.go#L9

In my PoV there could be two ways to support exporting checkpoints via the client api:
* Change `docker checkpoint create` to do it similar to `docker cp` (streaming the content through the api to the client).
* Add a new parameter to `docker export` such as `docker export --checkpoint`. (Same for `docker import` to have complementing actions).

In my humble opinion I would prefer the second implementation as it would feel more natural as a CLI user to first create a checkpoint as persisted state on the host (like we do with `docker create`) and afterwards use `docker export` and `docker import` to transfer state from the daemon to the client and vice-versa.

Any comments?

**Comment by cpuguy83** [comment]
We need to re-evaluate the design of these commands/API endpoints.
There's an open issue about it, but I'm currently on mobile.

**Comment by marc-christian-schulze** [comment]
Didn't find any ticket referring to an api re-design. Could you please link the ticket so we can check whether it's already work in progress.

**Comment by cpuguy83** [comment]
Evidently scattered across different (closed) issues.
I opened https://github.com/moby/moby/issues/36433
